### ISSUE TITLE
[Hold][$2000] Unread message shows as bold in the LHN but green line does not appear

### ISSUE THREAD

**Issue by kavimuru** [Expensify/App#15019]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
Two cases: 

First Case
1. Send a message when receiver is not logged in
2. log into the receiver account
3. Check the message

Second Case
1. Open two sessions one as User A and one as User B
2. Make sure User B is navigate to the chat with User A and save the URL
3. Close User B tab (or kill app)
4. Create message as User A on chat with User B and wait 10 seconds
5. Open tab (or app) as User B using the URL saved from the second step (on native use a deep link)
6. See that new marker is set correctly
7. Verify that the chat in the LHN is marked as read

## Expected Result:
Both in the LHN and in the chat shows New message indicator

## Actual Result:
No new message indicator inside the chat, only in the LHN it shows as bold as new message indicator

## Workaround:
unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?

- [ ] MacOS / Chrome / Safari

**Version Number:** 1.2.68-0
**Reproducible in staging?:** y
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** 

https://user-images.githubusercontent.com/43996225/217967717-d75e1ae5-6e6d-4fb6-a128-3e32eb77e51d.mp4

https://user-images.githubusercontent.com/43996225/217967732-8a8e86fb-0c3d-4f26-adf3-05a9ffa532c0.mov

**Expensify/Expensify Issue URL:**
**Issue reported by:** @gadhiyamanan 
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1675942126884169

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01649bcaf76128383a</li>
        <li>Upwork Job ID: 1623884459709763584</li>
        <li>Last Price Increase: 2023-02-17</li>
    </ul>
</details>

**Comment by MelvinBot** [comment]
Triggered auto assignment to @bfitzexpensify (`Bug`), see https://stackoverflow.com/c/expensify/questions/14418 for more details.

**Comment by MelvinBot** [comment]
### Bug0 Triage Checklist ([Main S/O](https://stackoverflowteams.com/c/expensify/questions/14418))

- [ ] This "bug" occurs on a supported platform (ensure `Platforms` in OP are ✅)
- [ ] This bug is not a duplicate report (check [E/App issues](https://github.com/Expensify/App/issues) and [#expensify-bugs](https://expensify.slack.com/archives/C049HHMV9SM))
    - If it is, comment with a link to the original report, close the issue and add any novel details to the original issue instead
- [ ] This bug is reproducible using the reproduction steps in the OP. [S/O](https://stackoverflowteams.com/c/expensify/questions/14418#:~:text=The%20bug%20is%20reproducible)
    - If the reproduction steps are clear and you're unable to reproduce the bug, check with the reporter and QA first, then close the issue.
    - If the reproduction steps aren't clear and you determine the correct steps, please update the OP.
- [ ] This issue is filled out as thoroughly and clearly as possible
    - Pay special attention to the title, results, platforms where the bug occurs, and if the bug happens on staging/production.
- [ ] I have reviewed and subscribed to the linked Slack conversation to ensure Slack/Github stay in sync

**Comment by MelvinBot** [comment]
Job added to Upwork: https://www.upwork.com/jobs/~01649bcaf76128383a

**Comment by MelvinBot** [comment]
Current assignee @bfitzexpensify is eligible for the External assigner, not assigning anyone new.

**Comment by MelvinBot** [comment]
Triggered auto assignment to Contributor-plus team member for initial proposal review - @thesahindia (`External`)

**Comment by MelvinBot** [comment]
Triggered auto assignment to @bondydaa (`External`), see https://stackoverflow.com/c/expensify/questions/7972 for more details.

**Comment by bfitzexpensify** [comment]
Just adding here, this is the green line in question that's missing:

<img width="1200" alt="image" src="https://user-images.githubusercontent.com/38236554/217992306-bdaf32df-a977-4aeb-b613-82fc9a284092.png">

**Comment by abdulrahuman5196** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
User logs in to his account and already has unread messages. If user goes to the unread chat 'new' message indicator doesn't show up.

### What is the root cause of that problem?
Background 1: When the user logs in, we fetch the list of all reports(this doesn't include report actions). Now when user goes to any chat we call openReport API to fetch all the details of the chat including report actions. This openReport API also updates the lastReadTime to current  time in the chat.
Background 2: The new indicator is shown based on internal state maintained in the ReportActionsView.js, it is initially set when the component is created and it gets updated based on user actions like message deletes, manual unreads. This is done to keep an internal state to show the new message indicator even if the last read time changes after chat is shown.

Root cause:
1) Now as per background1, when you open a unread chat after logging in we don't have report actions(empty list) and the 'newMarkerReportActionID' value is set as empty here - https://github.com/Expensify/App/blob/main/src/pages/home/report/ReportActionsView.js#L70. 
2) Now as per background2, we are parallely calling openReport API which fetches the list of report actions and also updates the lastReadTime to current. 
3) now when the information of latest report actions is propagated to ReportActionsView.js, it also has the lastReadTime to the latest Time as fetched from openReport API. So ReportActionsView.js takes the new lastReadTime and doesn't show the new message indicator

### What changes do you think we should make in order to solve the problem?
I think there is a core problem in having newMarkerReportActionID as state to show new indicator which represents the message id to show as new. But we end up doing lot of checks and complex logic to make the newMarkerReportActionID a proper value when the message gets deleted or new message is marked as unread. It seems like we are re-doing the old sequence number unread logic.

I propose instead of having newMarkerReportActionID as a state to show new indicator, have the 'lastReadTime' of a report as a state to show the new line indicator and new message button indicator as well.
To flow will be like this.
1) When user opens the chat from LHN, we store the 'lastReadTime' in a state, so that we would show the next message after the time with a 'new' message indicator and appropriate 'new message' button as per requirement.
2) We update this 'lastReadTime' only on user interactions like 'unread message', 'scrollToBottomAndMarkReportAsRead' and any other user action if required.
3) While displaying we will check which message is be shown the 'new' message indicator.

Advantages:
1) This way we don't have to worry if a message is deleted/available or not. The new message indicator will automatically shift to the next message when a new line indicator message is deleted. And it will fix this current issue of messages not being available at first place as well.
2) We can remove multiple checks and complex logics in keeping the newMarkerReportActionID up-to-date here - https://github.com/Expensify/App/blob/main/src/pages/home/report/ReportActionsView.js#L179
3) This will be in-sync with the new sequence number migration which is completed and we can start to rely on lastReadTime instead of hardcoded messageIds(seems to be like old sequence number implementation)
4) This will not impact any current behaviours.

This could solve these 3 bugs in one shot
https://github.com/Expensify/App/issues/13740
https://github.com/Expensify/App/issues/14649
https://github.com/Expensify/App/issues/15019

Only one disadvantage i see is that we would compute which message to show up as a new message while displaying, but it shouldn't be that complex computation as well as we only compare lastReadTime and action created time.

As this is not going to be an one line change there are multiple code-paths we can achieve the replacement of newMarkerReportActionID with lastReadTime state. One of best path at current codebase would be to maintain the state at reportScreen and propogate as we don't mount ReportActionsView(data won't be propagate to it when unmounted state). We can also do the same at other places but the core proposal is to remove the state of maintaining message ids and move to maintaining state of lastReadTime.

### What alternative solutions did you explore? (Optional)

We can add checks in the ReportActionsView.js to update the newMarkerReportActionID after the report is fully loaded, but as mentioned above we would end up adding more checks and complex logics similar to this in future to make newMarkerReportActionID up-to-date as per requirements. So the alternative seems to be a short term solution.

**Comment by s77rt** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Opening an unread report for the first time is missing the new marker indicator

### What is the root cause of that problem?
Opening any report will not only return the report actions but also set the last read time of that report to the current time. This is not necessary a bug. I'm considering this as a design decision and working on top of that.

To help understand the problem let's consider two unread reports:
- Report 1: Unread report actions are loaded into onyx (got them via pusher as we were already signed in)
- Report 2: Unread report actions are not loaded into onyx 

And let's consider the working flow after you click a report as follow:
1. Render report (and [set the new marker indicator](https://github.com/Expensify/App/blob/14ebe26b6af9c19b7debedb25bcbdf976cb57c5e/src/pages/home/report/ReportActionsView.js#L70) if eligible)
2. Make `OpenReport` API request (and update the last read time)

Rendering the first report will have the new marker indicator set since we have unread report actions in onyx. However that is not the case with the second report, in fact the second report will be rendered without the new messages (despite being shown on LHN) and only after `OpenReport` response will be rendered but at this point the last read time is also updated.

### What changes do you think we should make in order to solve the problem?
Fetch unread report actions before rendering the report. This can be achieved by having the server prefetch report actions for prefetched reports (or at least for prefetched unread reports). The report actions should be returned as a part of the `OpenApp` response (this needs to be done internally). I have created this [slack thread](https://expensify.slack.com/archives/C01GTK53T8Q/p1676040052829089) to get more feedback on that approach.

### What alternative solutions did you explore? (Optional)
None

**Comment by abdulrahuman5196** [comment]
@thesahindia Thank you. I have raised slack conversation here - https://expensify.slack.com/archives/C01GTK53T8Q/p1676267128690339

**Comment by abdulrahuman5196** [comment]
@thesahindia I have updated the proposal with clear information as per slack request - https://github.com/Expensify/App/issues/15019#issuecomment-1425434440
I have also added a alternative 2 proposal which should be simpler and can solve this particular issue alone if required.

**Comment by MelvinBot** [comment]
@bondydaa, @bfitzexpensify, @thesahindia Whoops! This issue is 2 days overdue. Let's get this updated quick!

**Comment by JmillsExpensify** [comment]
Coming from this issue, please make sure that the case from https://github.com/Expensify/App/issues/12598 is also addressed. I've also added the reproduction steps from that issue at the top of this one. Thanks!

**Comment by bondydaa** [comment]
i'm going OOO and don't want to be a blocker for progress so assigned to Puneet for now, I'll check back when I'm online next week again.

**Comment by MelvinBot** [comment]
@puneetlath, @bfitzexpensify, @thesahindia Uh oh! This issue is overdue by 2 days. Don't forget to update your issues!

**Comment by bondydaa** [comment]
removed `help wanted` and threw a hold on the title

**Comment by MelvinBot** [comment]
@puneetlath, @bfitzexpensify, @thesahindia Uh oh! This issue is overdue by 2 days. Don't forget to update your issues!

**Comment by puneetlath** [comment]
Ok, so what we're thinking is that the best path will be to take this to a design doc to lay out what exactly we think the ideal "new marker" behavior should be and then how to implement that. 

@bondydaa do you want to take that forward?

**Comment by bondydaa** [comment]
going to close this out in favor of centralizing discussion on https://github.com/Expensify/App/issues/15212.

**Comment by bondydaa** [comment]
whoops sorry I jumped the gun on closing, leaving open and on hold for now and dropping priority to monthly

**Comment by bondydaa** [comment]
tracking issue is still being worked on

**Comment by quinthar** [comment]
This is a regression right?

**Comment by bondydaa** [comment]
@quinthar this is being handled on this tracking issue https://github.com/Expensify/App/issues/15212

**Comment by bfitzexpensify** [comment]
Tracking issue continues to progress

**Comment by Victor-Nyagudi** [comment]
This issue also occurs after flagging a message as spam, inconsiderate, etc. 

https://github.com/Expensify/App/assets/79470910/85de219f-ac5d-4135-8720-6e7dd822e157

I thought I'd create a new bug, but it looked to similar to [the current issue](https://github.com/Expensify/App/issues/15019). 

There's another recently opened [issue where marking Concierge's message as unread doesn't work when it's the last message](https://github.com/Expensify/App/issues/20109) that could be worthy of being added to the [tracking issue](https://github.com/Expensify/App/issues/15212).

Thoughts, @bondydaa?

**Comment by bondydaa** [comment]
looks like design doc was reviewed and implementation has started per https://github.com/Expensify/App/issues/15212#issuecomment-1642013719

**Comment by bondydaa** [comment]
looks like the tracking issue is getting close 🎉 

I'm going to retest these steps now.

**Comment by bondydaa** [comment]
Hmm I'm not sure if these are the intended behaviors or not anymore but this is what I'm seeing:

Logging in, no new message indicator: https://github.com/Expensify/App/assets/4073354/0c0f764f-82d6-4291-9fc9-d17b6d2a2b1d

Even when if I leave the tab open, it looks like a new message shows up then gets marked as read quickly after: https://github.com/Expensify/App/assets/4073354/9d89865d-dce1-456a-a022-57b388f00b89

cc @MonilBhavsar

**Comment by bondydaa** [comment]
oh sorry looks like the 2nd thing is being tracked here https://github.com/Expensify/App/issues/23171

**Comment by bondydaa** [comment]
looks like https://github.com/Expensify/App/issues/23171 is still open

**Comment by bfitzexpensify** [comment]
Still on hold

**Comment by MonilBhavsar** [comment]
This should be fixed! Can we please retest and close this issue
